presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeAssignmentStats.java

```diff
@@ -20,43 +20,85 @@
-            assignmentCount.put(node, 0);
+        for (RemoteTask task : existingTasks) {
+            checkArgument(stageQueuedSplitInfo.put(task.getNodeId(), new PendingSplitInfo(task.getQueuedPartitionedSplitCount())) == null, "A single stage may not have multiple tasks running on the same node");
         }
 
-        for (RemoteTask task : existingTasks) {
-            checkArgument(queuedSplitCountByNode.put(task.getNodeId(), task.getQueuedPartitionedSplitCount()) == null, "A single stage may not have multiple tasks running on the same node");
+        // pre-populate the assignment counts with zeros
+        if (existingTasks.size() < nodeMapSize) {
+            Function<String, PendingSplitInfo> createEmptySplitInfo = (ignored) -> new PendingSplitInfo(0);
+            for (InternalNode node : nodeMap.getNodesByHostAndPort().values()) {
+                stageQueuedSplitInfo.computeIfAbsent(node.getNodeIdentifier(), createEmptySplitInfo);
+            }
         }
     }
 
     public int getTotalSplitCount(InternalNode node)
     {
-        return assignmentCount.getOrDefault(node, 0) + splitCountByNode.computeIfAbsent(node, nodeTaskMap::getPartitionedSplitsOnNode);
+        int nodeTotalSplits = nodeTotalSplitCount.computeIfAbsent(node, nodeTaskMap::getPartitionedSplitsOnNode);
+        PendingSplitInfo stageInfo = stageQueuedSplitInfo.get(node.getNodeIdentifier());
+        return nodeTotalSplits + (stageInfo == null ? 0 : stageInfo.getAssignedSplitCount());
     }
 
     public int getQueuedSplitCountForStage(InternalNode node)
     {
-        return queuedSplitCountByNode.getOrDefault(node.getNodeIdentifier(), 0) + assignmentCount.getOrDefault(node, 0);
+        PendingSplitInfo stageInfo = stageQueuedSplitInfo.get(node.getNodeIdentifier());
+        return stageInfo == null ? 0 : stageInfo.getQueuedSplitCount();
     }
 
     public void addAssignedSplit(InternalNode node)
     {
-        assignmentCount.merge(node, 1, (x, y) -> x + y);
+        String nodeId = node.getNodeIdentifier();
+        // Avoids the extra per-invocation lambda allocation of computeIfAbsent since assigning a split to an existing task more common than creating a new task
+        PendingSplitInfo stageInfo = stageQueuedSplitInfo.get(nodeId);
+        if (stageInfo == null) {
+            stageInfo = new PendingSplitInfo(0);
+            stageQueuedSplitInfo.put(nodeId, stageInfo);
+        }
+        stageInfo.addAssignedSplit();
+    }
+
+    private static final class PendingSplitInfo
+    {
+        private final int queuedSplitCount;
+        private int assignedSplits;
+
+        private PendingSplitInfo(int queuedSplitCount)
+        {
+            this.queuedSplitCount = queuedSplitCount;
+        }
+
+        public int getAssignedSplitCount()
+        {
+            return assignedSplits;
+        }
+
+        public int getQueuedSplitCount()
+        {
+            return queuedSplitCount + assignedSplits;
```

Comment:
It feels a bit weird to have `getQueuedSplitCount` return `queuedSplitCount + assignedSplits`, but I don't have a better name in mind.

Reply:
Yeah, the structure here is a little quirky (but is fundamentally the same as before, just with 1 fewer `HashMap`). Any split assigned within the current batch is "effectively queued" even though it hasn't made its way to the task yet. I couldn't think of a name that worked any better at making that piece clear.